packages/suite/src/views/wallet/send/components/Options/components/BitcoinOptions/components/CoinControl.tsx

```diff
@@ -0,0 +1,218 @@
+    margin-top: 20px;
+`;
+
+const Line = styled.div`
+    width: 100%;
+    height: 1px;
+    border-top: 1px solid ${({ theme }) => theme.STROKE_GREY};
+    margin: 12px 0 16px 0;
+`;
+
+interface Props {
+    close: () => void;
+}
+
+export const CoinControl = ({ close }: Props) => {
+    const [currentPage, setSelectedPage] = useState(1);
+
+    const { addToast, fetchTransactions } = useActions({
+        addToast: notificationActions.addToast,
+        fetchTransactions: fetchTransactionsThunk,
+    });
+
+    const { translationString } = useTranslation();
+
+    const {
+        account,
+        allUtxosSelected,
+        composedInputs,
+        dustUtxos,
+        errors,
+        getDefaultValue,
+        isCoinControl,
+        network,
+        outputs,
+        selectedUtxos,
+        spendableUtxos,
+        toggleCheckAllUtxos,
+        toggleCoinControl,
+    } = useSendFormContext();
+
+    const { areSatsUsed } = useBitcoinAmountUnit(account.symbol);
+
+    const inputs = isCoinControl ? selectedUtxos : composedInputs;
+
+    const getTotal = (amounts: number[]) =>
+        amounts.reduce((previous, current) => previous + current, 0);
+    const getFormattedAmount = (amount: number) =>
+        formatNetworkAmount(amount.toString(), account.symbol);
+
+    // calculate and format amounts
+    const totalInputs = getTotal(inputs.map(input => Number(input.amount)));
+    const totalOutputs = getTotal(
+        outputs.map((_, i) => Number(getDefaultValue(`outputs[${i}].amount`, ''))),
+    );
+    const totalOutputsInSats = areSatsUsed ? totalOutputs : totalOutputs * 10 ** network.decimals;
+    const missingToInput = totalOutputsInSats - totalInputs;
+    const formattedTotal = getFormattedAmount(totalInputs);
+    const formattedMissing = getFormattedAmount(missingToInput);
+    const isMissingToAmount = missingToInput > 0;
+    const isMissingVisible =
+        isCoinControl &&
+        (isMissingToAmount ||
+            !!errors.outputs?.some(
+                error =>
+                    ((error?.amount as TypedFieldError)?.message as ExtendedMessageDescriptor)
+                        ?.id === 'TR_NOT_ENOUGH_SELECTED',
+            ));
+    const missingToInputId = isMissingToAmount ? 'TR_MISSING_TO_INPUT' : 'TR_MISSING_TO_FEE';
+
+    // pagination
+    const totalItems = account.utxo?.length || 0;
+    const utxosPerPage = SETTINGS.TXS_PER_PAGE;
+    const showPagination = totalItems > utxosPerPage;
+    const lastSpendableUtxoIndex = currentPage * utxosPerPage;
+    const spendableUtxosOnPage = spendableUtxos.slice(
+        lastSpendableUtxoIndex - utxosPerPage,
+        lastSpendableUtxoIndex,
+    );
+    const lastDustUtxoIndex = currentPage * utxosPerPage - spendableUtxos.length;
+    const dustOnPage = dustUtxos.slice(
+        lastDustUtxoIndex - utxosPerPage,
+        lastDustUtxoIndex > 0 ? lastDustUtxoIndex : 0,
+    );
+
+    useEffect(() => {
+        try {
+            fetchTransactions({
```

Comment:
you forget effect cleanup this will be broken in React 18

Reply:
All effect are fired twice in React 18 in dev mode https://beta.reactjs.org/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development

Reply:
I believe firing twice in dev mode is not a problem. But maybe it could lead to some race conditions? `fetchTransaction` could be running multiple times with different params and there is no guarantee  on the order of completion. But we use it like this in [TransactionList](https://github.com/trezor/trezor-suite/blob/develop/packages/suite/src/views/wallet/transactions/components/TransactionList/index.tsx/#L71) already. 🤔

Reply:
Ah, that! I read about it too but didn't connect the dots. We would refactor it an use [this](https://redux-toolkit.js.org/api/createAsyncThunk#canceling-while-running). But it's applicable to so many existing actions in Suite :(

Reply:
It should work because transactions are already refactored to use RTK

Reply:
@Nodonisko @dahaca 1512d8dfbd138e1b08449fd97268849d1665a090 - what do you think?

Reply:
Seems reasonable, I am quite suprised how much code it added :D But it's probably because thunk itself is quite complicated